Add support for exporting Serato Tags

Holzhaus opened the pull request:
This adds basic support for exporting hotcues, BPM, beatgrid lock state and track color to Serato tags. Exporting the beatgrid is not supported at the moment, as I'm waiting for @hacksdump to finish #2961 first.

Props to @uklotzde for pointing me at the right location to retrieve the `StreamInfo`.

**Testing instructions:**

~~Build with this patch applied:~~ (not necessary anymore) 
```diff
$ git apply <<EOF
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 303f47ab0a..2120b8bdbe 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -922,7 +922,7 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
 )
 set_target_properties(mixxx-lib PROPERTIES AUTOMOC ON AUTOUIC ON)
 target_include_directories(mixxx-lib PUBLIC src "${CMAKE_CURRENT_BINARY_DIR}/src")
-target_compile_definitions(mixxx-lib PRIVATE SETTINGS_FILE="mixxx.cfg")
+target_compile_definitions(mixxx-lib PRIVATE SETTINGS_FILE="mixxx.cfg" __EXPORT_SERATO_MARKERS__)
 if(UNIX AND NOT APPLE)
   target_compile_definitions(mixxx-lib PRIVATE SETTINGS_PATH=".mixxx/")
 endif()
EOF
```
Set a few cue points, then restart Mixxx. Now you can export the metadata even when not loading the track onto a deck. Restart again, then reset all metadata, reimport it from file tags and your cue points should show up again.

**To Do:**
- [x] Rebase onto #3413 (after merging)
- [ ] ~~Add configuration options for Serato Tag export (always, don't overwrite, never).~~ On further thought this doesn't really make sense.
- [X] Get rid of the ifdef ~~(probably not in Mixxx 2.3 though)~~
- [ ] Test this with Serato

Holzhaus committed:
Track: Add basic support for exporting Serato Tags

uklotzde commented on src/sources/soundsourceproxy.cpp:
```diff
@@ -348,16 +348,26 @@ ExportTrackMetadataResult
-    if (pMetadataSource) {
-        return pTrack->exportMetadata(pMetadataSource);
-    } else {
+    if (!pSoundSource ||
```
We need to apply the same algorithm as used in openAudioSource(). This function has to be generalized to work for this case by omitting to wrap the result into AudioSourceTrackProxy.

uklotzde replied:
I will prepare a solution that can be integrated as a prerequisite.

Holzhaus replied:
Rebased on #3413.

uklotzde reviewed (changes requested):
Inconsistent opening of audio sources.

Holzhaus committed:
Preferences: Allow enabling/disabling Serato Metadata export

Be-ing commented:
What is the point of putting this behind an #ifdef if the user can toggle it at runtime in the preferences?

Holzhaus commented:
> What is the point of putting this behind an #ifdef if the user can toggle it at runtime in the preferences?

Feature freeze. Generating the Serato Metadats BLOBs already tested by the unit tests, but I'm not sure if we want to expose that feature to users without more testing yet. Also, the Beatgrid is *not* exported yet, because I'm waiting for @hacksdump to finish the new Beats implementation.

Be-ing commented:
That doesn't make sense to me. Either we have sufficient confidence in this feature to include it in 2.3 or it should go to main. I think it should be off by default even if it works perfectly. Writing nonstandard tags to files should be explicitly opt-in IMO.

Holzhaus committed:
Track: Clean up Serato Metadata export code

Holzhaus committed:
Serato Tags: Prevent data loss by skipping export if parsing failed

Holzhaus commented:
> That doesn't make sense to me. Either we have sufficient confidence in this feature to include it in 2.3 or it should go to main. I think it should be off by default even if it works perfectly. Writing nonstandard tags to files should be explicitly opt-in IMO.

I do have sufficient confidence. However, the feature is incomplete due to the missing beatgrid export. Not sure if we should enable it by default just yet. Since #2961 hasn't moved much since GSoC ended, I'll look into implementing the "Mixxx Beatgrid to Serato glue code" with the old API.

Holzhaus committed:
Track: Make Serato timing offset const

Holzhaus commented:
OK, I implemented support for exporting the beatgrid to Serato metadata in #3421 (for testing, you can use #3435). With that merged, I think we can remove the ifdefs. The checkbox label still clearly communicates that export is experimental.

uklotzde commented on src/preferences/dialog/dlgpreflibrary.cpp:
```diff
@@ -98,11 +102,6 @@ DlgPrefLibrary::DlgPrefLibrary(
 #endif
     builtInFormats->setText(builtInFormatsStr);
 
-    connect(checkBox_SyncTrackMetadataExport,
```
This got lost!?

uklotzde commented:
Apart from the finding the code is ok.

Be-ing commented:
So... thoughts about merging this? @uklotzde?

uklotzde commented:
Oh, I see there are still open TODOs? Does this reflect the current status?

Be-ing commented:
@Holzhaus do you want to finish this for 2.3?

Holzhaus committed:
Track: Remove #ifdefs around serato export code

Holzhaus committed:
Track: Add support for Serato BeatGrid export

Holzhaus commented:
During testing I encountered this debug assertion when exporting track metadata:

    DEBUG ASSERT: "0 == (samples % getChannelCount())" in function SINT mixxx::audio::SignalInfo::samples2frames(SINT) const at /home/jan/Projects/mixxx/src/audio/signalinfo.h:42

Didn't look into it further yet.

daschuer commented:
I can imagine that this originates from beats between frames. At one point we need to round to a full frame. Maybe this layer is missing at one point.

Holzhaus commented:
> I can imagine that this originates from beats between frames. At one point we need to round to a full frame. Maybe this layer is missing at one point.

Correct.

This line causes the issue: https://github.com/mixxxdj/mixxx/blob/main/src/track/serato/beatgrid.cpp#L450
I'm later calling samples2frames followed by frames2secs on it. Do we have a convenience function for doing the rounding? Alternatively, do we have a function to round samples2secs directly?

daschuer commented:
currentBeatPositionSamples must be a double and the static cast to SINT needs to be removed.

Holzhaus commented:
Thanks, I already took care of that locally. I also wanted to ensure correctness using test but I'm having some issues with creating beats test data: https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Serato.20cue.20import/near/229684033

Holzhaus commented:
I rebases this on top of https://github.com/mixxxdj/mixxx/pull/3688 which should be merged first.

Holzhaus commented:
I think this is ready to review, test and merge.

Be-ing commented:
Does not work for me:
```
Debug [Main]: GlobalTrackCache - Evicting track  TrackRef {"/home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac","/home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac",2640} Track(0x1edfe460)
Debug [Main]: SoundSourceProxy - SoundSourceProvider "Xiph.org libFLAC" created a SoundSource for file "file:///home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac" of type "flac"
Critical [Main]: DEBUG ASSERT: "streamInfo && streamInfo->isValid()" in function ExportTrackMetadataResult Track::exportMetadata(mixxx::MetadataSourcePointer, UserSettingsPointer) at ../src/track/track.cpp:1368
Warning [Main]: Track - Cannot write Serato metadata because stream info is not available: "/home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac"
Debug [Main]: TrackCollectionManager - Saving track "/home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac" in internal collection
Debug [Main]: TrackDAO: Saving track 2640 QFileInfo(/home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac)
Debug [Main]: TrackDAO: Updating track in database 2640 QFileInfo(/home/be/music/BUMBLE - 2017 - Tether Together/BUMBLE - Tether Together - 01 Story of a Story.flac)
Debug [Main]: SqlTransaction - Started new SQL database transaction on "MIXXX-1"
Debug [Main]: SqlTransaction - Committed SQL database transaction on "MIXXX-1"
Debug [Main]: BaseTrackCache(0xf1d59d0) updateIndexWithQuery took 0 ms
Debug [Main]: GlobalTrackCache - Deleting Track(0x1edfe460)
Debug [Main]: 2842 ms deleting DlgPreferences
```

Be-ing commented on src/preferences/dialog/dlgpreflibrarydlg.ui:
```diff
@@ -138,10 +138,16 @@
+      <item row="1" column="0">
+       <widget class="QCheckBox" name="checkBox_SeratoMetadataExport">
+        <property name="text">
+         <string>Write Serato Metadata to files (experimental/potentially makes files unusable in Serato!)</string>
```
```suggestion
         <string>Write Serato Metadata to files</string>
```
I suggest replacing the ambiguous disclaimer with a confirmation dialog when the user enables the preference that says something like "Writing Serato metadata to files is experimental. It is recommended that you back up your music files before enabling it. Are you sure you want to enable this experimental feature?"

Holzhaus committed:
DlgPrefLibrary: Display warning QQMessageBox when enabling Serato Export

Be-ing commented on src/preferences/dialog/dlgpreflibrary.cpp:
```diff
@@ -330,6 +334,23 @@ void DlgPrefLibrary::slotRelocateDir() {
+                    QStringLiteral(
+                            "Exporting Serato Metadata from Mixxx is highly "
+                            "experimental. There is no official documentation "
+                            "of the format. Files with Serato metadata written "
+                            "by Mixxx might crash Serato DJ, therefore caution "
```
I do not think we should scare users by talking about Serato crashing if we have no reports of this actually happening.

Be-ing replied:
IMO it is more important to advise users to backup their music files so they can revert to a backup in case Mixxx messes up their tags.

Be-ing replied:
So yes, we should warn users, but we don't need to be dramatic about it.

Holzhaus replied:
I don't think you need a backup. You can easily remove the Serato tags, e.g. `eyeD3 --remove--all-objects somefile.mp3`.

Holzhaus replied:
Except if we overwrite existing Serato metadata...

Be-ing replied:
Recommending a backup is not a bad idea in any case.

Holzhaus replied:
> Except if we overwrite existing Serato metadata...

On second thought that shouldn't happen because we check if parsing worked fine, and skip writing if there were any errors. But I'll add backups anyway.

Be-ing commented on src/preferences/dialog/dlgpreflibrary.cpp:
```diff
@@ -327,11 +334,30 @@ void DlgPrefLibrary::slotRelocateDir() {
+        if (QMessageBox::warning(this,
+                    QStringLiteral("Serato Metadata Export"),
+                    QStringLiteral(
+                            "Exporting Serato Metadata from Mixxx is highly "
```
```suggestion
                            "Exporting Serato Metadata from Mixxx is "
```

Holzhaus commented:
@uklotzde Is it not guaranteed that we have valid signal info during export?

GorgiAstro commented:
Hi all, I came across this pull request on the forum while investigating how to synchronize cue points between two computers.

I built mixxx on Holzhaus' branch serato-tag-export@3b8507081084a39f026f1fd38f61a7d6efbdc94e

When exporting tags to metadata for all crates, it works for MP3 files, but for FLAC files I get the same issue as Be-ing. Here's an extract of my log file (containing one successful MP3 export and one unsuccessful FLAC export):

> Debug [Main]: Track - Old metadata (imported) TrackMetadata{ streamInfo: StreamInfo{ signalInfo: SignalInfo{ channelCount: 2 sampleRate: 44100 Hz } bitrate: 280 kbps duration: 404000000000 ns } trackInfo: TrackInfo{ artist: "David Joseph" bpm: 0 comment: "" composer: "" genre: "Pop" grouping: "" key: "" replayGain: ratio = 0 / peak = -1 seratoTags: SeratoTags title: "You Can't Hide (Your Love From Me) (12'' Version)" trackNumber: "2" trackTotal: "10" year: "2005" } albumInfo: AlbumInfo{ artist: "" title: "Grand 12-Inches 2 (Cd 3)" } }
> Debug [Main]: Track - New metadata (modified) TrackMetadata{ streamInfo: StreamInfo{ signalInfo: SignalInfo{ channelCount: 2 sampleRate: 44100 Hz } bitrate: 280 kbps duration: 404297142857 ns } trackInfo: TrackInfo{ artist: "David Joseph" bpm: 115.892 comment: "" composer: "" genre: "Pop" grouping: "" key: "F♯m" replayGain: ratio = 0.552146 / peak = -1 seratoTags: SeratoTags title: "You Can't Hide (Your Love From Me) (12'' Version)" trackNumber: "2" trackTotal: "10" year: "2005" } albumInfo: AlbumInfo{ artist: "" title: "Grand 12-Inches 2 (Cd 3)" } }
> Debug [Main]: MetadataSourceTagLib - Exporting track metadata into file "/media/yzokras/ABSOLUTE_UNIT/owncloud/redacted/Grand 12inches 2 [V0]/Cd3/02 - David Joseph - You Cant Hide (Your Love From Me) (12 Version).mp3" with type 3
> Debug [Main]: TrackCollectionManager - Saving track "/media/yzokras/ABSOLUTE_UNIT/owncloud/redacted/Grand 12inches 2 [V0]/Cd3/02 - David Joseph - You Cant Hide (Your Love From Me) (12 Version).mp3" in internal collection
> Debug [Main]: Successfully deserialized BeatGrid
> Debug [Main]: Successfully deserialized KeyMap
> Debug [Main]: BaseTrackCache(0x55ec03e3f440) updateIndexWithQuery took 0 ms
> Warning [Main]: Track - Cannot write Serato metadata because stream info is not available: "/media/yzokras/ABSOLUTE_UNIT/owncloud/redacted/Simon Dunmore - Glitterbox - Love Is The Message (2017) [FLAC]/12 Eleanor Mills - Mr Right (The Right mix).flac"
> Debug [Main]: TrackCollectionManager - Saving track "/media/yzokras/ABSOLUTE_UNIT/owncloud/redacted/Simon Dunmore - Glitterbox - Love Is The Message (2017) [FLAC]/12 Eleanor Mills - Mr Right (The Right mix).flac" in internal collection

Holzhaus commented:
Thanks for testing! Does this happen for all FLAC files, or just that specific one?

GorgiAstro commented:
For all FLAC files that I've tested (a dozen).

Holzhaus commented:
Hmm, I guess this is a bug in our FLAC code, not a bug in the Serato tag export code. @uklotzde Any idea what causes this?

uklotzde commented:
> Hmm, I guess this is a bug in our FLAC code, not a bug in the Serato tag export code. @uklotzde Any idea what causes this?

StreamInfo::isValid() is too restrictive. No bitrate is set.

StreamInfo{ signalInfo: SignalInfo{ channelCount: 2 sampleRate: 44100 Hz } bitrate: 0 kbps duration: 331480000000 ns }

uklotzde commented:
I suggest to explicitly check for only those properties that are actually required, i.e. SignalInfo::isValid() and a non-empty duration.

uklotzde commented:
The FLAC decoder doesn't report a bitrate. This is only reported by TagLib.

Holzhaus committed:
SeratoTags: Use SignalInfo/Duration instead of StreamInfo

For FLAC files, the StreamInfo might be invalid because the bitrate is
invalid, because it is read by TagLib, not the sound source.  We don't
actually need the Bitrate for exporting Serato Tags, but the signal info
alone does not suffice either, because we also need the duration to
serialize the beatgrid. Hence, instead of using the StreamInfo wrapper
(which also has the irrelevant Bitrate member) we use SignalInfo and
Duration directly.

GorgiAstro commented:
I tested commit 7f07185, now I manage to export serato tags to FLAC metadata, thanks!

Holzhaus committed:
DlgPrefLibrary: Improve Serato Metadata Export warning message

uklotzde commented:
I suggest to append " (experimental)" to the preference label. This disclaimer should be visible before users actually decide to enable the option. Users are used to click away pop-up dialogs without reading them ;)

Holzhaus committed:
DlgPrefLibrary: Add "(experimental)" to Serato metadata export checkbox

uklotzde commented:
I have added the 2.3.0 label. Exporting both cues and beat grid as file tags would be awesome and complete the round trip.

uklotzde commented on src/preferences/dialog/dlgpreflibrarydlg.ui:
```diff
@@ -138,10 +138,16 @@
+      <item row="1" column="0">
+       <widget class="QCheckBox" name="checkBox_SeratoMetadataExport">
+        <property name="text">
+         <string>Write Serato Metadata to files</string>
```
Write Serato Metadata to files (experimental)

uklotzde replied:
Did @Be-ing ask you to remove it earlier? I disagree. Such a warning should not only appear in a popup dialog, hidden in a paragraph of text.

uklotzde reviewed (approved):
My remark regarding the label text doesn't prevent merging this PR.

As a user without a deeper understanding of the consequences I expect to see this warning before enabling the option. This would also increase the probability that users are willing to read the text in the popup dialog.

Holzhaus commented:
@Be-ing is the warning message okay now? If so, please merge.

Be-ing commented on src/preferences/dialog/dlgpreflibrary.cpp:
```diff
@@ -327,11 +334,32 @@ void DlgPrefLibrary::slotRelocateDir() {
+                            "lost and files with Serato metadata written by "
+                            "Mixxx could potentially crash Serato DJ, "
+                            "therefore caution is advised and backups are "
+                            "recommended. Do you really want to enable this "
```
```suggestion
                            "recommended. Are you sure you want to enable this "
```

Holzhaus replied:
Done.

Holzhaus committed:
DlgPrefLibrary: Rephrase Serato Metadata Export message

Co-authored-by: Be <be.0@gmx.com>

uklotzde commented:
@Holzhaus @Be-ing Merge?

Holzhaus commented:
Yes.
